File: neuroflow_core_slib/layer_order_comparer.cpp
#include "stdafx.h"
#include "layer_order_comparer.h"
#include "layer.h"
#include "layer_connections.h"

USING

bool layer_order_comparer::operator()(const layer_ptr& layer1, const layer_ptr& layer2)
{
    return compare(layer1, layer2) < 0 ? true : false;
}

int layer_order_comparer::compare(const layer_ptr& layer1, const layer_ptr& layer2)
{
    auto key = make_pair(layer1, layer2);
    
    auto result = results.find(key);
    if (result != results.end()) return result->second;

    auto rKey = make_pair(layer2, layer1);

    result = results.find(rKey);
    if (result != results.end()) return -result->second;

    int ret;
    if (is_below(layer1, layer2)) ret = -1;
    else if (is_below(layer2, layer1)) ret = 1;
    else ret = 0;

    results.insert(make_pair(key, ret));
    return ret;
}

bool layer_order_comparer::is_below(const layer_ptr& layer1, const layer_ptr& layer2)
{
    for (auto& output : layer1->output_connections().connected_layers(flow_direction::one_way | flow_direction::two_way))
    {
        if (output == layer2) return true;
    }
    for (auto& output : layer1->output_connections().connected_layers(flow_direction::one_way | flow_direction::two_way))
    {
        if (is_below(output, layer2)) return true;
    }
    return false;
}

File: _vault/prev_version_win/Neuroflow/NeuralNetworks/LayerOrderComparer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Neuroflow.NeuralNetworks
{
    internal sealed class LayerOrderComparer : IComparer<Layer>
    {
        Dictionary<Tuple<Layer, Layer>, int> results = new Dictionary<Tuple<Layer, Layer>, int>();

        public int Compare(Layer x, Layer y)
        {
            var key = Tuple.Create(x, y);
            int result;

            if (results.TryGetValue(key, out result)) return result;

            var rKey = Tuple.Create(y, x);

            if (results.TryGetValue(rKey, out result)) return -result;

            if (IsBelow(x, y)) result = -1;
            else if (IsBelow(y, x)) result = 1;
            else result = 0;
            results.Add(key, result);
            return result;
        }

        private static bool IsBelow(Layer x, Layer y)
        {
            foreach (var output in x.OutputConnections.GetConnectedLayers(FlowDirection.OneWay | FlowDirection.TwoWay))
            {
                if (output == y) return true;
            }
            foreach (var output in x.OutputConnections.GetConnectedLayers(FlowDirection.OneWay | FlowDirection.TwoWay))
            {
                if (IsBelow(output, y)) return true;
            }
            return false;
        }
    }
}
